File: solutions/12100.py
from typing import List, Tuple

class Board:
    def __init__(self, size: int, board: List[List[int]]):
        self.size = size
        self.board = board

    def move(self, direction):
        if direction == (0, 1):
            for column in reversed(range(self.size)):
                top = self.size - 1
                for row in range(self.size):
                    if self.board[row][column] != 0:
                        self.board[top][column] = self.board[row][column]
                        self.board[row][column] = 0
                        top -= 1
        if direction == (1, 0):
            for row in range(self.size):
                top = self.size - 1
                for column in reversed(range(self.size)):
                    if self.board[row][column] != 0:
                        self.board[row][top] = self.board[row][column]
                        self.board[row][column] = 0
                        top -= 1
        if direction == (0, -1):
            for column in range(self.size):
                top = 0
                for row in range(self.size):
                    if self.board[row][column] != 0:
                        self.board[top][column] = self.board[row][column]
                        self.board[row][column] = 0
                        top += 1
        if direction == (-1, 0):
            for row in range(self.size):
                top = 0
                for column in range(self.size):
                    if self.board[row][column] != 0:
                        self.board[row][top] = self.board[row][column]
                        self.board[row][column] = 0
                        top += 1
                

    def get_piece(self, pos: Tuple[int, int]) -> int:
        return self.board[pos[0]][pos[1]]
    
    def set_piece(self, pos: Tuple[int, int], value: int) -> None:
        self.board[pos[0]][pos[1]] = value

    def merge(self, direction):
        lookup = {
            (1, 0): ((0, self.size - 1), (0, self.size)),
            (0, 1): ((0, self.size), (0, self.size - 1)),
            (-1, 0): ((1, self.size), (0, self.size)),
            (0, -1): ((0, self.size), (1, self.size))
        }
        ranges = lookup[direction]
        for i in range(ranges[0][0], ranges[0][1]):
            for j in range(ranges[1][0], ranges[1][1]):
                if self.get_piece((i, j)) == self.get_piece( (i + direction[0], j + direction[1]) ):
                    self.set_piece((i, j), 2 * self.get_piece((i, j)))
    
    def play(self, direction):
        self.move(direction)
        self.merge(direction)
        self.move(direction)


board = [
        [2, 0, 0, 0],
        [2, 2, 0, 0],
        [2, 0, 0, 0],
        [0, 0, 0, 0]
        ]

b = Board(4, board)

print(b.board)

while True:
    a = input()
    if a =='h':
        b.play((-1, 0))
    if a =='j':
        b.play((0, -1))
    if a =='k':
        b.play((0, 1))
    if a =='l':
        b.play((1, 0))

    print(b.board)

File: solutions/12100.cpp
#include "../modules/Matrix.h"
#include "../modules/Math.h"
#include "../modules/FastIO2.h"
#include "../modules/FastIO.h"

uint N;

Int2 starting_point(Int2 dir) {
	if (dir == Int2(1, 0))
		return Int2(0, 0);
	if (dir == Int2(0, 1))
		return Int2(N - 1, 0);
	if (dir == Int2(-1, 0))
		return Int2(N - 1, N - 1);
	if (dir == Int2(0, -1))
		return Int2(0, N - 1);

	__builtin_unreachable();
}

void compress(Matrix<int>& board, Int2 dir) {
	Int2 search_dir = dir * -1;

	Int2 start = starting_point(search_dir);

	for (uint column = 0; column < N; column++) {
		Int2 column_start = start + search_dir.rotate(Math::pi / 2) * column;
		Int2 last_available = column_start;

		for (uint row = 0; row < N; row++) {
			Int2 index = column_start + search_dir * row;

			if (board[index] == 0) continue;

			int value = board[index];
			board[index] = 0;
			board[last_available] = value;

			last_available += search_dir;
		}
	}
}

void merge(Matrix<int>& board, Int2 dir) {
	Int2 search_dir = dir * -1;

	Int2 start = starting_point(search_dir);

	for (uint column = 0; column < N; column++) {
		Int2 column_start = start + search_dir.rotate(Math::pi / 2) * column;

		for (uint row = 0; row < N - 1; row++) {
			Int2 index = column_start + search_dir * row;

			if (board[index] == 0) continue;

			Int2 next_index = index + search_dir;

			if (board[index] == board[next_index]) {
				board[index] *= 2;
				board[next_index] = 0;
			}
		}
	}
}

void play(Matrix<int>& board, Int2 dir) {
	compress(board, dir);
	merge(board, dir);
	compress(board, dir);
}

int search(Matrix<int>& board, uint plays_left) {
	if (plays_left == 0) {
		int result = 0;

		for (Int2 index: board.bounds())
			result = std::max(result, board[index]);

		return result;
	}

	static Int2 moves[4] = {
		Int2(0, 1),
		Int2(1, 0),
		Int2(0, -1),
		Int2(-1, 0)
	};

	int result = 0;

	for (Int2 move: moves) {
		Matrix<int> copy = board;
		play(copy, move);
		result = std::max(result, search(copy, plays_left - 1));
	}

	return result;
}

int main() {
	FastIO();
	Fast::cin >> N;
	Matrix<int> board(N, N);
	Fast::cin >> board;

	std::cout << search(board, 5);
}
